src/runtime/virtcontainers/qemu.go

```diff
@@ -1304,13 +1304,23 @@ func (q *qemu) hotplugVFIODevice(device *config.VFIODev, op operation) (err erro
 			case config.VFIODeviceNormalType:
 				return q.qmpMonitorCh.qmp.ExecuteVFIODeviceAdd(q.qmpMonitorCh.ctx, devID, device.BDF, device.Bus, romFile)
 			case config.VFIODeviceMediatedType:
+				if utils.IsAPVFIOMediatedDevice(device.SysfsDev) {
+					return q.qmpMonitorCh.qmp.ExecuteAPVFIOMediatedDeviceAdd(q.qmpMonitorCh.ctx, device.SysfsDev)
+				}
 				return q.qmpMonitorCh.qmp.ExecutePCIVFIOMediatedDeviceAdd(q.qmpMonitorCh.ctx, devID, device.SysfsDev, "", device.Bus, romFile)
 			default:
 				return fmt.Errorf("Incorrect VFIO device type found")
 			}
 		}
 
-		addr, bridge, err := q.arch.addDeviceToBridge(devID, types.PCI)
+		var bridgeType types.Type
+		if machinneType == QemuCCWVirtio {
+			bridgeType = types.CCW
+		} else {
+			bridgeType = types.PCI
+		}
+
+		addr, bridge, err := q.arch.addDeviceToBridge(devID, bridgeType)
```

Comment:
does s390x support PCI or PCIE bridges? if not, could you implement this function (`addDeviceToBridge`)  in qemu_s390x.go ?  that way the above `if..else` condition won't be needed here

Reply:
It supports PCI for certain devices, but it is not used in Kata.

Reply:
Here: https://wiki.qemu.org/Documentation/Platforms/S390X#A_note_on_PCI_support

Reply:
@devimc are you suggesting to ignore the second param in `addDeviceToBridge` in qemu_s390x and add set locally the bridge type?

Reply:
@devimc adjusted, were you thinking something like this?

Reply:
kind of..  I'll implement  it this way because I consider that the address format `(%04x)` and bridge type (ccw) is s390x specific, what do you think? let me know if this works for you

```patch
diff --git a/src/runtime/virtcontainers/qemu_arch_base.go b/src/runtime/virtcontainers/qemu_arch_base.go
index 259907e8..363bc3fd 100644
--- a/src/runtime/virtcontainers/qemu_arch_base.go
+++ b/src/runtime/virtcontainers/qemu_arch_base.go
@@ -712,30 +712,34 @@ func (q *qemuArchBase) setIgnoreSharedMemoryMigrationCaps(ctx context.Context, q
 }
 
 func (q *qemuArchBase) addDeviceToBridge(ID string, t types.Type) (string, types.Bridge, error) {
+	addr, b, err := genericAddDeviceToBridge(q.Bridges, ID, t)
+	if err != nil {
+		return "", b, err
+	}
+
+	return fmt.Sprintf("%02x", addr), b, nil
+}
+
+func genericAddDeviceToBridge(bridges []types.Bridge, ID string, t types.Type) (uint32, types.Bridge, error) {
 	var err error
 	var addr uint32
 
-	if len(q.Bridges) == 0 {
-		return "", types.Bridge{}, errors.New("failed to get available address from bridges")
+	if len(bridges) == 0 {
+		return 0, types.Bridge{}, errors.New("failed to get available address from bridges")
 	}
 
 	// looking for an empty address in the bridges
-	for _, b := range q.Bridges {
+	for _, b := range bridges {
 		if t != b.Type {
 			continue
 		}
 		addr, err = b.AddDevice(ID)
 		if err == nil {
-			switch t {
-			case types.CCW:
-				return fmt.Sprintf("%04x", addr), b, nil
-			case types.PCI, types.PCIE:
-				return fmt.Sprintf("%02x", addr), b, nil
-			}
+			return addr, b, nil
 		}
 	}
 
-	return "", types.Bridge{}, fmt.Errorf("no more bridge slots available")
+	return 0, types.Bridge{}, fmt.Errorf("no more bridge slots available")
 }
 
 func (q *qemuArchBase) removeDeviceFromBridge(ID string) error {
diff --git a/src/runtime/virtcontainers/qemu_s390x.go b/src/runtime/virtcontainers/qemu_s390x.go
index df890a7d..7617d6de 100644
--- a/src/runtime/virtcontainers/qemu_s390x.go
+++ b/src/runtime/virtcontainers/qemu_s390x.go
@@ -269,3 +269,12 @@ func (q *qemuS390x) appendVSock(devices []govmmQemu.Device, vsock types.VSock) (
 func (q *qemuS390x) appendIOMMU(devices []govmmQemu.Device) ([]govmmQemu.Device, error) {
 	return devices, fmt.Errorf("S390x does not support appending a vIOMMU")
 }
+
+func (q *qemuS390x) addDeviceToBridge(ID string, t types.Type) (string, types.Bridge, error) {
+	addr, b, err := genericAddDeviceToBridge(q.Bridges, ID, t)
+	if err != nil {
+		return "", b, err
+	}
+
+	return fmt.Sprintf("%04x", addr), b, nil
+}
-- 
2.26.2
```

Reply:
@devimc works for me, updated like this. I put you as a co-author (woo fancy GitHub features) if that's okay.